File: lib/ledger_dsl/transaction.rb
# frozen_string_literal: true
# typed: strict

require 'sorbet-runtime'
require 'date'

require_relative 'status'
require_relative 'posting'
require_relative 'comment'
require_relative 'date'

# DSL for creating financial journals
module LedgerDSL
  # Represents a financial transaction
  class Transaction
    extend T::Sig

    sig do
      params(
        date: ToDate,
        status: Status,
        code: T.nilable(String),
        payee: T.nilable(String),
        note: T.nilable(String),
        tags: String
      ).void
    end
    def initialize(date, status: Status::Unmarked, code: nil, payee: nil, note: nil, **tags)
      @entries = T.let([], T::Array[T.any(Posting, Comment)])
      @date = T.let(LedgerDSL.date_from(date), Date)
      @status = T.let(status, Status)
      @code = T.let(code, T.nilable(String))
      @payee = T.let(payee, T.nilable(String))
      @note = T.let(note, T.nilable(String))
      @tags = T.let(tags, T::Hash[Symbol, String])
    end

    sig { params(date: T.nilable(ToDate)).returns(Date) }
    def date(date = nil) = date.nil? ? @date : @date = LedgerDSL.date_from(date)

    sig { params(status: T.nilable(Status)).returns(Status) }
    def status(status = nil) = status.nil? ? @status : @status = status

    sig { params(code: T.nilable(String)).returns(T.nilable(String)) }
    def code(code = (reader = true; nil)) = reader ? @code : @code = code

    sig { params(payee: T.nilable(String)).returns(T.nilable(String)) }
    def payee(payee = (reader = true; nil)) = reader ? @payee : @payee = payee

    sig { params(note: T.nilable(String)).returns(T.nilable(String)) }
    def note(note = (reader = true; nil)) = reader ? @note : @note = note

    sig { returns(T::Hash[Symbol, String]) }
    attr_reader :tags

    sig do
      params(
        account: String,
        status: Status,
        amount: T.nilable(ToCostedAmount),
        balance: T.nilable(ToCostedAmount),
        comment: T.nilable(ToComment),
        date: T.nilable(ToDate),
        date2: T.nilable(ToDate),
        tags: String,
        blk: T.nilable(T.proc.bind(Posting).params(p: Posting).void)
      ).returns(Posting)
    end
    def posting(
      account, status: Status::Unmarked, amount: nil, balance: nil,
      comment: nil, date: nil, date2: nil, **tags, &blk
    )
      posting = Posting.new(account, status: status, amount: amount, balance: balance, comment: comment, date: date,
                                     date2: date2, **tags)
      posting.instance_exec(posting, &blk) if block_given?
      @entries << posting
      posting
    end
    alias po posting

    # Create a new comment in the transaction
    sig { params(comment: ToComment).returns(Comment) }
    def comment(comment)
      comment = Comment.from(comment)
      @entries << Comment.from(comment)
      comment
    end
    alias co comment

    sig { params(_indent: Integer).returns(String) }
    def to_ledger(_indent = 0)
      o = @date.to_s
      o += ' !' if @status == Status::Pending
      o += ' *' if @status == Status::Cleared
      o += " (#{@code})" if @code
      o += " #{@payee}" if @payee
      o += " | #{@note}" if @note

      tags_str = @tags.map { |k, v| "#{k}:#{v}" }.join(', ')
      o += " ; #{tags_str}" unless tags_str.empty?

      o += "\n"
      o += @entries.map { |entry| entry.to_ledger(4) }.join("\n")

      o
    end
  end
end

File: src/entities/transaction.ts
import { formatDate } from "../ledger/date";
import { Comment } from "./comment";
import { Posting } from "./posting";
import { Status } from "./status";

/**
 * Represents a transaction.
 */
export class Transaction {
  #date: Date;
  #date2: Date | null = null;
  public status: Status = Status.Unmarked;
  public code: string | null = null;
  public payee: string | null = null;
  public note: string | null = null;
  public tags: Map<string, string> = new Map();
  #entries: Array<Posting | Comment>;

  constructor(
    date: Date | string,
    status?: Status,
    code?: string | null,
    payee?: string | null,
    note?: string | null,
    date2?: Date | null,
    tags?: Record<string, string>,
  );
  constructor(date: Date | string, entries: Array<Posting | Comment>);
  constructor(
    date: Date | string,
    statusOrEntries: Status | Array<Posting | Comment> = Status.Unmarked,
    code: string | null = null,
    payee: string | null = null,
    note: string | null = null,
    date2: Date | null = null,
    tags: Record<string, string> = {},
  ) {
    this.#date = new Date(date);
    this.date2 = date2;
    this.code = code;
    this.payee = payee;
    this.note = note;
    this.tags = new Map(Object.entries(tags));

    if (Array.isArray(statusOrEntries)) {
      this.status = Status.Unmarked;
      this.#entries = [...statusOrEntries];
      return;
    }
    this.status = statusOrEntries;
    this.#entries = [];
  }

  /**
   * Sets the date of the transaction.
   */
  public set date(date: Date) {
    this.#date = new Date(date);
  }

  /**
   * Gets the date of the transaction.
   */
  public get date(): Date {
    return this.#date;
  }

  /**
   * Gets the secondary date associated with the transaction.
   */
  public get date2(): Date | null {
    return this.#date2;
  }

  /**
   * Sets the secondary date associated with the transaction.
   */
  public set date2(date2: Date | string | null) {
    this.#date2 = date2 === null ? null : new Date(date2);
  }

  /**
   * Add an entry to the transaction.
   */
  protected addEntry(entry: Posting | Comment): void {
    this.#entries.push(entry);
  }

  /**
   * Converts the transaction to a ledger-formatted string.
   *
   * @returns The formatted ledger string.
   */
  public toLedger(): string {
    let output = formatDate(this.#date);

    if (this.#date2) {
      output += `=${formatDate(this.#date2)}`;
    }

    if (this.status === Status.Pending) {
      output += "! ";
    } else if (this.status === Status.Cleared) {
      output += "* ";
    }

    if (this.code) {
      output += ` (${this.code})`;
    }
    if (this.payee) {
      output += ` ${this.payee}`;
    }
    if (this.note) {
      output += ` | ${this.note}`;
    }

    const tagsStr = Array.from(this.tags.entries())
      .map(([key, value]) => `${key}:${value}`)
      .join(", ");
    if (tagsStr !== "") {
      output += ` ; ${tagsStr}`;
    }

    output += "\n";
    output += this.#entries.map((entry) => entry.toLedger(4)).join("\n");

    return output;
  }
}
